### ISSUE TITLE
Reassign ColocatedTasks assigned to non VLJ support team members

### ISSUE THREAD

**Issue by hschallhorn** [department-of-veterans-affairs/caseflow#12173]
VSCLHALL and ADJEWILH were added to the Colocated org in Caseflow although they do not work for the VLJ support team at the board. This has left them with automatically assigned `ColocatedTask`s due to the fact that tasks assigned to the colocated org get automatically assigned to anyone in the organization.

## Blocked By
- [x] These non team members should be removed from the Colocated org in the teams management page by one of the VLJ Support team admins. (@stwards11 should notify when this has happened)

## AC
- [ ] All `ColocatedTask`s previously assigned to `VSCLHALL` and `ADJEWILH` are reassigned to other members of the colocated staff.

**Comment by kevmo** [comment]
1 | |||||||||
2 | |||
3 | 
5 |
8 |

1.

**Comment by hschallhorn** [comment]
See [this tech spec](https://github.com/department-of-veterans-affairs/caseflow/pull/12197/files#diff-33a91cd5134bf0c5a4522e456e900342R38) for some guidance. Buddy up on this!

**Comment by hschallhorn** [comment]
- [ ] Start with assigned and in_progress (active) colocated tasks
```ruby
VSCLHALL = User.find_by(css_id: "VSCLHALL")
active_tasks = ColocatedTask.active.where(assigned_to: VSCLHALL)

# Let's test on one task
task = active_tasks.first
puts task.reload.appeal.structure_render(:id, :assigned_to_id, :status)
# LegacyAppeal 884327
# ├── RootTask 305239, 2, assigned
# └── ScheduleHearingColocatedTask 308530, 23, on_hold
#     └── ScheduleHearingColocatedTask 308531, 10425, assigned

task.parent.assigned_to.next_assignee
# <User id: 7798, ...
# This is who we would assume the next task will be assigned to

# Alright, let's test the reassigning!
task.reassign({ assigned_to_type: User.name, assigned_to_id: task.parent.assigned_to.next_assignee.id }, task.assigned_by)
puts task.reload.appeal.structure_render(:id, :assigned_to_id, :status)
# LegacyAppeal 884327
# ├── RootTask 305239, 2, assigned
# └── ScheduleHearingColocatedTask 308530, 23, on_hold
#     ├── ScheduleHearingColocatedTask 308531, 10425, cancelled
#     └── ScheduleHearingColocatedTask 456129, 7798, assigned

# Looks good, let's test that next assignee was incremented
task.parent.assigned_to.next_assignee
# => <User id: 9967, ...
# ✅ 

# Let's take on the rest!
active_tasks = ColocatedTask.active.where(assigned_to: VSCLHALL)

# Let's make sure all parent tasks are assigned to colocated
parent_tasks = Task.where(id: active_tasks.pluck(:parent_id))
parent_tasks.group(:assigned_to_id).count
# => {23=>202}
# Awesome, we can use the next_assignee function to reassign them

# First, let's batch these bad boys
active_colocated_user_count = Colocated.singleton.users.count
# => 21
num_tasks_to_assign_each = active_tasks.count / active_colocated_user_count
# => 9

active_tasks.in_groups_of(num_tasks_to_assign_each).each do |group|
  next_assignee_id = Colocated.singleton.next_assignee.id
  group.each do |task|
    task.reassign({ assigned_to_type: User.name, assigned_to_id: next_assignee_id }, task.assigned_by) if task
  end
end
```

Test on one, confirm new assignee, confirm next_assignee increments 
- [ ] Handle on_hold tasks
- [ ] Track down other task types that were assigned to Colocated team

**Comment by hschallhorn** [comment]
@stwards11, as of this moment, just under 550 of the tasks in these two user's queues have been reassigned evenly amongst the 21 VLJ support team members. That's about 26 new tasks that should show up in these users queues if you want to make Maria aware that this has happened. We still have fewer than 55 tasks to reassign, so this is not completely done, but the bulk of it is.

```ruby
 Task.open.where(assigned_to: VSCLHALL).group(:type).count
# => {"StayedAppealColocatedTask"=>1, "ExtensionColocatedTask"=>10, "OtherColocatedTask"=>1, "PowerOfAttorneyRelatedMailTask"=>1, "EvidenceOrArgumentMailTask"=>15, "ExtensionRequestMailTask"=>2}
Task.open.where(assigned_to: ADJEWILH).group(:type).count
# => {"ExtensionColocatedTask"=>9, "OtherColocatedTask"=>2, "EvidenceOrArgumentMailTask"=>14}

**Comment by hschallhorn** [comment]
Reassigning active non `ColocatedTask` as other task types can be assigned to the colocated team. We'll be excluding `EvidenceOrArgumentMailTask`s for now


```ruby
user = User.find_by(css_id: "VSCLHALL")
task_ids_assigned_to_colocated = Task.open.where(assigned_to: Colocated.singleton).pluck(:id)
active_tasks = Task.active.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated).where.not(type: EvidenceOrArgumentMailTask.name)
active_tasks.count
# => 18

# As there are fewer tasks to reassign than colocated team members, theres nothing to gain from batching
active_tasks.each do |task|
  next_assignee = Colocated.singleton.next_assignee
  task.reassign({ assigned_to: next_assignee }, task.assigned_by)
end

# There should be no more active tasks assigned to VSCLHALL that was assigned to the colocated team
Task.active.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated).where.not(type: EvidenceOrArgumentMailTask.name).count
# => 0

pp Task.open.where(assigned_to: user).group(:type, :status).count
# {["ExtensionColocatedTask", "on_hold"]=>10,
#   ["OtherColocatedTask", "on_hold"]=>1,
#   ["StayedAppealColocatedTask", "on_hold"]=>1,
#   ["EvidenceOrArgumentMailTask", "assigned"]=>15,
#   ["TimedHoldTask", "assigned"]=>12}

# Great! The only active tasks in this user's queue are `EvidenceOrArgumentMailTask`s
# Let's repeat all this for ADJEWILH
user = User.find_by(css_id: "ADJEWILH")
task_ids_assigned_to_colocated = Task.open.where(assigned_to: Colocated.singleton).pluck(:id)
active_tasks = Task.active.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated).where.not(type: EvidenceOrArgumentMailTask.name)
active_tasks.count
# => 0
pp Task.open.where(assigned_to: user).group(:type, :status).count
# {["TimedHoldTask", "assigned"]=>11,
#   ["ExtensionColocatedTask", "on_hold"]=>9,
#   ["OtherColocatedTask", "on_hold"]=>2,
#   ["EvidenceOrArgumentMailTask", "assigned"]=>6}

# Looks like this user doesn't have any more active tasks to reassign!
# ✅ 
```

**Comment by hschallhorn** [comment]
All that's left is to handle the on hold tasks in these user's queues! Down to a final 23 tasks!
``` ruby
user = User.find_by(css_id: "VSCLHALL")
task_ids_assigned_to_colocated = Task.open.where(assigned_to: Colocated.singleton).pluck(:id)
on_hold_tasks = Task.on_hold.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated)
on_hold_tasks.group(:type).count
# => {"StayedAppealColocatedTask"=>1, "ExtensionColocatedTask"=>10, "OtherColocatedTask"=>1}

pp on_hold_tasks.order(:type).pluck(:type, :instructions)
# [["ExtensionColocatedTask",
#   ["Veteran asked for 90-days to submit evidence.  Please hold in abeyance until October 9, 2019 and send a letter to the Veteran and his representative noting the same.  ",
#    "Counsel has requested file to be held in abeyance until 10.09.19 to allow rep the full 90-days from Board docket letter dated 7.11.19"]],
#  ["ExtensionColocatedTask",
#   ["Attorney requested 90 day extension August 2019. ",
#    "Rep has requested file to be held for the full 90 days from Board docket date 7.30.19"]],
#  ["ExtensionColocatedTask",
#   ["On Aug 1, 2019, VA received a request for an extension of 60 days.  Please respond to the extension request and, if necessary, put the file in abeyance for the necessary length of time.  Thanks.",
#    "Grant letter to rep file will be held for 60-days"]],
#  ["ExtensionColocatedTask",
#   ["Attorney sent letter on 9/9/19 in response to 90- day extension and asked for add'l 30 days with the 90 days",
#    "Rep has requested file to be placed on hold for the full 90-days  from Board cert. letter dated 9.9.19"]],
#  ["ExtensionColocatedTask",
#   ["Please hold this in abeyance until 90 days from the docketing letter (July 29, 2019) and then return the case to me.  Thank you.",
#    "Counsel request file to be placed in abeyance for full 90 days"]],
#  ["ExtensionColocatedTask",
#   ["Please draft letter for judges signature to grant extension request.",
#    "Grant letter sent to judge for review and signature"]],
#  ["ExtensionColocatedTask",
#   ["In August 2019, the appellant's attorney requested that we wait the full 90 days from the 7/24/19 notification letter and an additional 30 days after that to allow for evidence and argument submission.  Please notify the attorney that the request is granted, and he has until 1/24/20 to submit evidence and argument.",
#    "Counsel has requested file to be held the full 90-days"]],
#  ["ExtensionColocatedTask",
#   ["Please see 8/14/2019 letter from the Veteran's representative requesting a 90 day extension for submission of a brief and additional evidence",
#    "Rep request 90 days from Board docket date 8.6.19"]],
#  ["ExtensionColocatedTask",
#   ["On 08/19/2019, the Veteran's attorney requested a 60-day extension. Please prepare a motion granting the requested extension and submit it to VLJ Paul Sorisio for his signature.",
#    "Grant letter sent to judge for signature and review"]],
#  ["ExtensionColocatedTask",
#   ["In correspondence received on August 22, 2019 the Veteran's attorney asked for an extension, specifically a full 90 days from the date of the August 13, 2019 Board letter, to submit additional evidence in reference to the claim.  Please  get an extension approval letter and place this case on hold until at least November 13, 2019.  Thanks.",
#    "Grant letter sent to judge for signature and review"]],
#  ["OtherColocatedTask",
#   ["In a July 2019 statement, the Veteran's representative requested the full 90 days before adjudication of this case. Please place the case in abeyance for the appropriate period. Thanks!",
#    "Granted extension"]],
#  ["StayedAppealColocatedTask",
#   ["This case is stayed in accordance with the Secretary's July 1, 2019 direction to stay adjudication of all cases which may be affected by the Blue Water Navy Vietnam Veterans Act of 2019 (the Act), until the effective date of the Act, January 1, 2020.  ",
#    "Blue Water Stay is in effect"]]]

user = User.find_by(css_id: "ADJEWILH")
task_ids_assigned_to_colocated = Task.open.where(assigned_to: Colocated.singleton).pluck(:id)
on_hold_tasks = Task.on_hold.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated)
on_hold_tasks.group(:type).count
# => {"ExtensionColocatedTask"=>9, "OtherColocatedTask"=>2}

pp on_hold_tasks.order(:type).pluck(:type, :instructions)
# [["ExtensionColocatedTask",
#   ["Veteran requested Board wait the full 90 days from June 5, 2019 to submit additional evidence for claims on appeal.",
#    "Rep request  full 90 days"]],
#  ["ExtensionColocatedTask",
#   ["Good afternoon - in correspondence received September 13, 2019, the Veteran requested that the Board wait the full 90 days from the August 28, 2019, letter before adjudicating the claims on appeal.  Please return to me when this time has expired.  Thank you,  Jeff",
#    "Rep has requested full 90-days from Board certification."]],
#  ["ExtensionColocatedTask",
#   ["Veteran's attorney J. Michael Woods requests a 90 day extension plus 30 days of processing from the date of his 8/26/2019 letter before a decision is issued.  Attorney states that he is waiting on favorable medical evidence.  Please hold case in abeyance until time lapses and issue letter granting extension. ",
#    "Drafted grant letter sent to judge for signature and review."]],
#  ["ExtensionColocatedTask",
#   ["hold for 90 days to receive additional evidence.",
#    "Per counsel request to hold for the full 90-days to allow submission of new and additional evidence"]],
#  ["ExtensionColocatedTask",
#   ["Please grant extension request and hold in abeyance for the requested time. Then return this file to me; this was a prior OT case from Greg Deemer.",
#    "Rep has requested file to be held for the full 90-days after docket date 7.30.19"]],
#  ["ExtensionColocatedTask",
#   ["Veteran's attorney requested extension of 90 days (plus 30 days for VA mail processing) in order to submit additional evidence and a brief. See letter dated 6/26/2019. ",
#    "Rep has requested full 90-days"]],
#  ["ExtensionColocatedTask",
#   ["Please prepare response to motion for extension of 90 days from August 19, 2019 (but not the extra 30 days) and hold in abeyance until the end of the extension.  ",
#    "Grant letter sent to judge for review and signature."]],
#  ["ExtensionColocatedTask",
#   ["Please process the extension request that was received on July 24th.  Thank you.",
#    "The grant letter has been sent to judge for signature and review."]],
#  ["ExtensionColocatedTask",
#   ["See email to Admin Action Team 1 on 8/29/19 - letter granting extension request should be prepared for Judge Vince Chiappetta's signature and case held in abeyance until 90 days have elapsed.",
#    "Rep has requested the full 90-days from docket letter dated 7.29.19"]],
#  ["OtherColocatedTask",
#   ["Veteran requests claim be held in abeyance for 120 days beginning August 7, 2019.",
#    "Rep has requested file to be held for 90-days plus 30-days mailing purposes"]],
#  ["OtherColocatedTask",
#   ["The Veteran's representative submitted a July 2019 letter in which he requested a full 90 days (plus an additional 30 days for VA mail processing) prior to adjudicating this case (see Third Party Correspondence with receipt date of 7/30/19). Please place the case in abeyance for the appropriate period. Thanks!",
#    "Rep has requested the full 90 days plus 30 days for mail purposes."]]]
```

For every task, it would appear that the user was correctly putting these tasks on hold. I would suggest that we keep these children on hold tasks. Thankfully, the [reassign function](https://github.com/department-of-veterans-affairs/caseflow/blob/e1a5d717fb51a34aeac10f276fc86a895266969f/app/models/task.rb#L479) handles that!
```ruby
user = User.find_by(css_id: "VSCLHALL")
task_ids_assigned_to_colocated = Task.open.where(assigned_to: Colocated.singleton).pluck(:id)
on_hold_tasks = Task.on_hold.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated)

# We'll start with one just to make sure
task  = on_hold_tasks.first
puts task.appeal.structure_render(:status, :assigned_to_id)
# LegacyAppeal 850379 [status, assigned_to_id]
# ├── RootTask assigned, 2
# └── ExtensionColocatedTask on_hold, 23
#     └── ExtensionColocatedTask on_hold, 10425
#         └── TimedHoldTask assigned, 10425

next_assignee_id = Colocated.singleton.next_assignee.id
task.reassign({ assigned_to_type: User.name, assigned_to_id: next_assignee_id }, task.assigned_by)

puts task.reload.appeal.structure_render(:id: status, :assigned_to_id)
# LegacyAppeal 850379 [id, status, assigned_to_id]
# ├── RootTask 296387, assigned, 2
# └── ExtensionColocatedTask 316319, on_hold, 23
#     ├── ExtensionColocatedTask 316320, cancelled, 10425
#     │   └── TimedHoldTask 357761, cancelled, 10425
#     └── ExtensionColocatedTask 465830, assigned, 1885

# Uh oh. Unfortunately, if we cancel a task, the timed hold task child will also be cancelled and not moved to the new task.
# Let's fix this case and whip up our own reassign function
Task.find(357761).update!(closed_at: nil, status: Constants.TASK_STATUSES.assigned, parent_id: 465830)
Task.find(465830).update!(status: Constants.TASK_STATUSES.on_hold)
puts task.reload.appeal.structure_render(:id: status, :assigned_to_id)
# LegacyAppeal 850379 [id, status, assigned_to_id]
# ├── RootTask 296387, assigned, 2
# └── ExtensionColocatedTask 316319, on_hold, 23
#     ├── ExtensionColocatedTask 316320, cancelled, 10425
#     └── ExtensionColocatedTask 465830, on_hold, 1885
#         └── TimedHoldTask 357761, assigned, 10425

# MUUUUUCH better
# Alright, let's put together a reassign function that should work with child TimedHoldTasks

def reassign(old_task, reassign_params, current_user)
  sibling = old_task.dup.tap do |task|
    task.assigned_by_id = old_task.class.child_assigned_by_id(old_task.parent, current_user)
    task.assigned_to = old_task.class.child_task_assignee(old_task.parent, reassign_params)
    task.instructions = [old_task.instructions, reassign_params[:instructions]].flatten
    task.status = Constants.TASK_STATUSES.assigned
    task.save!
  end

  old_task.children.open.each { |task| task.update!(parent_id: sibling.id) }
  sibling.update!(status: Constants.TASK_STATUSES.on_hold)
  old_task.update!(status: Constants.TASK_STATUSES.cancelled)

  [sibling, old_task, sibling.children].flatten
end

# Let's test on a single task again!
on_hold_tasks = Task.on_hold.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated)
task  = on_hold_tasks.first
puts task.appeal.structure_render(:status, :assigned_to_id)
# LegacyAppeal 898727 [status, assigned_to_id]
# ├── RootTask assigned, 2
# └── ExtensionColocatedTask on_hold, 23
#     └── ExtensionColocatedTask on_hold, 10425
#         └── TimedHoldTask assigned, 10425

next_assignee_id = Colocated.singleton.next_assignee.id
reassign(task, { assigned_to_type: User.name, assigned_to_id: next_assignee_id }, task.assigned_by)
puts task.appeal.structure_render(:status, :assigned_to_id)
# LegacyAppeal 898727 [status, assigned_to_id]
# ├── RootTask assigned, 2
# └── ExtensionColocatedTask on_hold, 23
#     ├── ExtensionColocatedTask cancelled, 10425
#     └── ExtensionColocatedTask on_hold, 2014
#         └── TimedHoldTask assigned, 10425

# Great! let's run this on all the on hold tasks!
on_hold_tasks = Task.on_hold.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated)
on_hold_tasks.each do |task|
  next_assignee_id = Colocated.singleton.next_assignee.id
  reassign(task, { assigned_to_type: User.name, assigned_to_id: next_assignee_id }, task.assigned_by)
end

Task.open.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated).count
# => 0
# All tasks have been reassigned from VSCLHALL!

# Let's wrap this up with ADJEWILH
user = User.find_by(css_id: "ADJEWILH")
task_ids_assigned_to_colocated = Task.open.where(assigned_to: Colocated.singleton).pluck(:id)
on_hold_tasks = Task.on_hold.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated)
on_hold_tasks.each do |task|
  next_assignee_id = Colocated.singleton.next_assignee.id
  reassign(task, { assigned_to_type: User.name, assigned_to_id: next_assignee_id }, task.assigned_by)
end
Task.open.where(assigned_to: user, parent_id: task_ids_assigned_to_colocated).count
# => 0
# All tasks have been reassigned from ADJEWILH!
# ✅✅✅✅✅✅✅
```
 @stwards11 All tasks assigned to these two users have been reassigned to the rest of the colocated team! Closing this ticket!
